### PR TITLE
CRYP-46: forgot password reset

### PR DESCRIPTION
**Pull request by Andre-Ibrahim**
IMG_2434
IMG_2428
IMG_2429
IMG_2433

### CODE REVIEW

File: packages/api/src/services/authentication.service.ts [hunk 1]
```diff
@@ -37,6 +41,27 @@ export class AuthenticationService {
         return this.signToken(user);
     }
 
+    async ForgotPassword(forgotPasswordReq: ForgotPasswordRequest): Promise<void> {
```
Comment (jason-gerard) [inline packages/api/src/services/authentication.service.ts:44]: Casing

```suggestion
    async forgotPassword(forgotPasswordReq: ForgotPasswordRequest): Promise<void> {
```

File: packages/api/src/services/authentication.service.ts [hunk 2]
```diff
@@ -37,6 +41,27 @@ export class AuthenticationService {
...
+
+        if (!user) {
+            throw new ForbiddenException();
+        }
+
+        const token = this.signToken(user);
+        this.forgotPasswordService.sendForgotPassword(user, token);
+    }
+
+    async ResetPassword(resetPasswordReq: ResetPasswordRequest): Promise<void> {
```
Comment (jason-gerard) [inline packages/api/src/services/authentication.service.ts:55]:
```suggestion
    async resetPassword(resetPasswordReq: ResetPasswordRequest): Promise<void> {
```

File: packages/api/src/services/authentication.service.ts [hunk 3]
```diff
@@ -37,6 +41,27 @@ export class AuthenticationService {
         return this.signToken(user);
     }
 
+    async ForgotPassword(forgotPasswordReq: ForgotPasswordRequest): Promise<void> {
+        const user = await this.usersService.findOne(forgotPasswordReq.email);
+
+        if (!user) {
+            throw new ForbiddenException();
```
Comment (jason-gerard) [inline packages/api/src/services/authentication.service.ts:48]: Probably a 404 here

File: packages/api/src/services/users.service.ts [hunk 4]
```diff
@@ -75,6 +75,13 @@ export class UsersService {
         return this.userRepository.save(user);
     }
 
+    async updatePassword(userId: number, newPassword: string): Promise<User> {
+        const user = await this.findOneById(userId);
+
+        user.password = await this.authService.encode(newPassword);
+        return this.userRepository.save(user);
+    }
```
Comment (jason-gerard) [inline packages/api/src/services/users.service.ts:83]: I think you should just use the update user service here

File: packages/client/screens/SignInScreen.tsx [hunk 5]
```diff
@@ -83,6 +84,13 @@ export default function SignInScreen() {
                             />
                             <FormControl.ErrorMessage>{errors.password}</FormControl.ErrorMessage>
                         </FormControl>
+                        <Link
+                            onPress={() => navigation.navigation.navigate("ResetPasswordEmailScreen")}
+                            _text={{ color: "darkBlue.500", fontWeight: "400" }}
+                            isUnderlined={false}
+                        >
+                            forgot Password?
```
Comment (jason-gerard) [inline packages/client/screens/SignInScreen.tsx:92]:
```suggestion
                            Forgot Password?
```

File: packages/common/src/utils/notifications/forgot_password_notification_service.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import { Injectable } from "@nestjs/common";
+import { ConfigService } from "@nestjs/config";
+import { User } from "@cryptify/common/src/domain/entities/user";
+import { EmailNotificationStrategy } from "./email_notification_strategy";
+import { JwtToken } from "../../domain/jwt_token";
+
+@Injectable()
+export class ForgotPasswordService {
```
Comment (jason-gerard) [inline packages/common/src/utils/notifications/forgot_password_notification_service.ts:8]: Current name doesnt make sense, here is a better one

```suggestion
export class AuthNotificationService {
```

File: packages/common/src/utils/notifications/forgot_password_notification_service.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    constructor(
+        private readonly emailNotificationStrategy: EmailNotificationStrategy,
+        private readonly configService: ConfigService,
+    ) {}
+
+    async sendForgotPassword(user: User, token: JwtToken): Promise<void> {
```
Comment (jason-gerard) [inline packages/common/src/utils/notifications/forgot_password_notification_service.ts:14]:
```suggestion
    async sendForgotPasswordEmail(user: User, token: JwtToken): Promise<void> {
```

File: packages/common/src/utils/notifications/forgot_password_notification_service.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        const appName = this.configService.get<string>("APP_NAME");
+        const appIP = this.configService.get<string>("APP_URL");
+        const appPort = this.configService.get<string>("APP_PORT");
+
+        const uri = `${appName}://${appIP}:${appPort}/--/forgot-password/${token.accessToken}`;
+
+        console.log(user);
```
Comment (jason-gerard) [inline packages/common/src/utils/notifications/forgot_password_notification_service.ts:21]:
```suggestion
```

File: packages/client/types.tsx [hunk 9]
```diff
@@ -213,6 +218,14 @@ export type GuestStackParamList = {
...
     SignUpNotificationsScreen: SignUpNotificationsScreenProps;
+    ResetPasswordEmailScreen: undefined;
+    CreateNewPasswordScreen: CreateNewPasswordScreenProps;
+    ResetPasswordSuccessScreen: undefined;
+    ResetPasswordFailureScreen: undefined;
+};
+
+export type CreateNewPasswordScreenProps = {
+    token: string;
 };
```
Comment (jason-gerard) [inline packages/client/types.tsx:229]: Move this above with the rest of the type defs

File: packages/client/screens/reset-password/createNewPasswordScreen.tsx [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+    const accessToken = navigation.route.params.token;
+
+    type formValues = ResetPasswordRequest & { confirmNewPassword: string };
+
+    async function onSubmitCreateNewPassword(values: formValues, formikHelpers: FormikHelpers<formValues>) {
+        try {
+            await authGateway.resetPassword({ token: accessToken, password: values.password });
+
+            console.log(navigation.navigation);
```
Comment (jason-gerard) [inline packages/client/screens/reset-password/createNewPasswordScreen.tsx:28]: extra

File: packages/api/src/services/authentication.service.ts [hunk 11]
```diff
@@ -37,6 +41,27 @@ export class AuthenticationService {
...
+        }
+
+        const token = this.signToken(user);
+        this.forgotPasswordService.sendForgotPassword(user, token);
```
Comment (jason-gerard) [inline packages/api/src/services/authentication.service.ts:52]: You probably want to await the notification here

File: packages/api/src/services/authentication.service.ts [hunk 12]
```diff
@@ -37,6 +41,27 @@ export class AuthenticationService {
...
+    }
+
+    async ResetPassword(resetPasswordReq: ResetPasswordRequest): Promise<void> {
+        const object = this.jwtService.decode(resetPasswordReq.token);
```
Comment (jason-gerard) [inline packages/api/src/services/authentication.service.ts:56]:
```suggestion
        const token = this.jwtService.decode(resetPasswordReq.token);
```

**Review by jason-gerard — APPROVED** [review]
Pretty good overall

File: packages/client/screens/reset-password/resetPasswordEmailScreen.tsx [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+        </View>
+    );
+}
+
+const styles = StyleSheet.create({
+    container: {
+        flex: 1,
+        alignItems: "center",
+        justifyContent: "center",
+        padding: 20,
```
Comment (domsec) [inline packages/client/screens/reset-password/resetPasswordEmailScreen.tsx:77]: No need to add padding to the top and bottom of the view. Vertical padding should be 15 and not 20.

```suggestion
        paddingVertical: 15,
```

File: packages/client/screens/reset-password/reset-password-sates/resetPasswordSuccessScreen.tsx [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+import { StyleSheet } from "react-native";
+import { FontAwesomeIcon } from "@fortawesome/react-native-fontawesome";
+import { farEnvelope } from "../../../components/icons/regular/farEnvelope";
+import { openInbox } from "react-native-email-link";
+
+export default function ResetPassworSuccessScreen() {
+    return (
+        <View style={styles.container}>
+            <VStack style={styles.content}>
+                <FontAwesomeIcon icon={farEnvelope} color={"#0077E6"} style={styles.envelopeIcon} size={100} />
```
Comment (domsec) [inline packages/client/screens/reset-password/reset-password-sates/resetPasswordSuccessScreen.tsx:13]: As previously mentioned, this should be a light font weight envelope icon with size 96
```suggestion
                <FontAwesomeIcon icon={farEnvelope} color={"#0077E6"} style={styles.envelopeIcon} size={96} />
```

File: packages/client/screens/reset-password/reset-password-sates/resetPasswordSuccessScreen.tsx [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+                <Text>Did not receive the email? Check your spam filter, or</Text>
+                <Link>try another email address</Link>
+            </VStack>
+        </View>
+    );
+}
+
+const styles = StyleSheet.create({
+    container: {
+        paddingHorizontal: 20,
```
Comment (domsec) [inline packages/client/screens/reset-password/reset-password-sates/resetPasswordSuccessScreen.tsx:33]:
```suggestion
        paddingHorizontal: 15,
```

File: packages/client/screens/reset-password/reset-password-sates/resetPasswordSuccessScreen.tsx [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+        justifyContent: "center",
+    },
+    content: {
+        flex: 1,
+        alignItems: "center",
+        marginTop: 50,
+    },
+    footer: {
+        alignItems: "center",
+        marginBottom: 100,
```
Comment (domsec) [inline packages/client/screens/reset-password/reset-password-sates/resetPasswordSuccessScreen.tsx:44]:
```suggestion
        marginBottom: 15,
```

File: packages/client/screens/reset-password/reset-password-sates/resetPasswordSuccessScreen.tsx [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+        justifyContent: "center",
+    },
+    envelopeIcon: {
+        marginTop: 20,
+        marginBottom: 20,
```
Comment (domsec) [inline packages/client/screens/reset-password/reset-password-sates/resetPasswordSuccessScreen.tsx:49]:
```suggestion
        marginBottom: 30,
```

File: packages/client/screens/reset-password/reset-password-sates/resetPasswordSuccessScreen.tsx [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+import { farEnvelope } from "../../../components/icons/regular/farEnvelope";
+import { openInbox } from "react-native-email-link";
+
+export default function ResetPassworSuccessScreen() {
+    return (
+        <View style={styles.container}>
+            <VStack style={styles.content}>
+                <FontAwesomeIcon icon={farEnvelope} color={"#0077E6"} style={styles.envelopeIcon} size={100} />
+                <Text size={"title1"}>Check Your Email</Text>
+                <Text style={{ paddingHorizontal: 40, marginTop: 10, textAlign: "center" }}>
```
Comment (domsec) [inline packages/client/screens/reset-password/reset-password-sates/resetPasswordSuccessScreen.tsx:15]:
```suggestion
                <Text style={{ paddingHorizontal: 40, marginTop: 15, textAlign: "center" }}>
```

File: packages/client/screens/reset-password/reset-password-sates/resetPasswordSuccessScreen.tsx [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+                    We have sent password recovery instructions to your email.
+                </Text>
+            </VStack>
+
+            <VStack style={styles.footer}>
+                <Button onPress={() => openInbox()} style={{ width: "100%", marginBottom: 10 }}>
```
Comment (domsec) [inline packages/client/screens/reset-password/reset-password-sates/resetPasswordSuccessScreen.tsx:21]:
```suggestion
                <Button onPress={() => openInbox()} style={{ width: "100%", marginBottom: 20 }}>
```

File: packages/client/screens/reset-password/reset-password-sates/resetPasswordSuccessScreen.tsx [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+import { FontAwesomeIcon } from "@fortawesome/react-native-fontawesome";
+import { farEnvelope } from "../../../components/icons/regular/farEnvelope";
+import { openInbox } from "react-native-email-link";
+
+export default function ResetPassworSuccessScreen() {
+    return (
+        <View style={styles.container}>
+            <VStack style={styles.content}>
+                <FontAwesomeIcon icon={farEnvelope} color={"#0077E6"} style={styles.envelopeIcon} size={100} />
+                <Text size={"title1"}>Check Your Email</Text>
```
Comment (domsec) [inline packages/client/screens/reset-password/reset-password-sates/resetPasswordSuccessScreen.tsx:14]:
```suggestion
                <Text size={"title3"} fontWeight={"semibold"}>Check Your Email</Text>
```

File: packages/client/screens/reset-password/reset-password-sates/resetPasswordSuccessScreen.tsx [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+                    We have sent password recovery instructions to your email.
+                </Text>
+            </VStack>
+
+            <VStack style={styles.footer}>
+                <Button onPress={() => openInbox()} style={{ width: "100%", marginBottom: 10 }}>
+                    Open email app
+                </Button>
+                <Text>Did not receive the email? Check your spam filter, or</Text>
+                <Link>try another email address</Link>
```
Comment (domsec) [inline packages/client/screens/reset-password/reset-password-sates/resetPasswordSuccessScreen.tsx:25]: Add proper styling
```suggestion
                <Link _text={{ color: "darkBlue.500", fontWeight: "semibold", size: "callout" }}
                            isUnderlined={false}>try another email address</Link>
```

File: packages/client/screens/reset-password/reset-password-sates/resetPasswordSuccessScreen.tsx [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+                <Text style={{ paddingHorizontal: 40, marginTop: 10, textAlign: "center" }}>
+                    We have sent password recovery instructions to your email.
+                </Text>
+            </VStack>
+
+            <VStack style={styles.footer}>
+                <Button onPress={() => openInbox()} style={{ width: "100%", marginBottom: 10 }}>
+                    Open email app
+                </Button>
+                <Text>Did not receive the email? Check your spam filter, or</Text>
```
Comment (domsec) [inline packages/client/screens/reset-password/reset-password-sates/resetPasswordSuccessScreen.tsx:24]:
```suggestion
                <Text size={"callout"}>Did not receive the email? Check your spam filter, or</Text>
```

File: packages/client/screens/reset-password/reset-password-sates/resetPasswordFailureScreen.tsx
Comment (domsec) [inline packages/client/screens/reset-password/reset-password-sates/resetPasswordFailureScreen.tsx:1]: To save more PR comments, please look at the comments in the `resetPasswordFailureScreen.tsx`. All comments apply to this file as well.

File: packages/client/screens/reset-password/createNewPasswordScreen.tsx [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+
+    const initialValues = {
+        token: accessToken,
+        password: "",
+        confirmNewPassword: "",
+    };
+
+    return (
+        <View style={styles.container}>
+            <Text size={"title3"} fontWeight={"semibold"}>
```
Comment (domsec) [inline packages/client/screens/reset-password/createNewPasswordScreen.tsx:45]:
```suggestion
            <Text size={"title1"} fontWeight={"semibold"}>
```

**Review by narroarrow — APPROVED** [review]
Well done

File: packages/client/screens/reset-password/createNewPasswordScreen.tsx [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+                Reset your password
```
Comment (domsec) [inline packages/client/screens/reset-password/createNewPasswordScreen.tsx:46]: Incorrect Title
```suggestion
                Create new password
```

File: packages/client/screens/reset-password/createNewPasswordScreen.tsx [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+            </Text>
+
+            <Text>
+                Enter the email address associated with your account and we'll send you instructions to reset your
+                password.
+            </Text>
```
Comment (domsec) [inline packages/client/screens/reset-password/createNewPasswordScreen.tsx:52]: Incorrect text
```suggestion
            <Text size={"callout"} style={{ marginTop: 20, textAlign: "center" }}>
Your new password must be different from your previously used password.
            </Text>
```

File: packages/client/screens/reset-password/createNewPasswordScreen.tsx [hunk 26]
```diff
@@ -0,0 +1,122 @@
...
+        </View>
+    );
+}
+
+const styles = StyleSheet.create({
+    container: {
+        flex: 1,
+        alignItems: "center",
+        justifyContent: "center",
+        padding: 20,
```
Comment (domsec) [inline packages/client/screens/reset-password/createNewPasswordScreen.tsx:116]:
```suggestion
        verticalPadding: 15,
```

File: packages/client/screens/reset-password/createNewPasswordScreen.tsx [hunk 27]
```diff
@@ -0,0 +1,122 @@
...
+                Enter the email address associated with your account and we'll send you instructions to reset your
+                password.
+            </Text>
+            <Formik
+                initialValues={initialValues}
+                validationSchema={resetPasswordSchema}
+                onSubmit={onSubmitCreateNewPassword}
+            >
+                {({ values, errors, touched, handleChange, submitForm }) => (
+                    <VStack space={4} marginTop={5} style={{ width: "100%" }}>
```
Comment (domsec) [inline packages/client/screens/reset-password/createNewPasswordScreen.tsx:59]:
```suggestion
                    <VStack space={"13px"} marginTop={"35px"} style={{ width: "100%" }}>
```

File: packages/client/screens/reset-password/createNewPasswordScreen.tsx [hunk 28]
```diff
@@ -0,0 +1,122 @@
...
+                                        />
+                                    </Pressable>
+                                }
+                                onChangeText={handleChange("confirmNewPassword")}
+                                placeholder="Confirm new password"
+                                testID="ConfirmNewPassword"
+                            />
+                            <FormControl.ErrorMessage>{errors.confirmNewPassword}</FormControl.ErrorMessage>
+                        </FormControl>
+                        <Button
```
Comment (domsec) [inline packages/client/screens/reset-password/createNewPasswordScreen.tsx:98]:
```suggestion
                        <Button style={{ marginTop: 7 }}
```

File: packages/client/screens/reset-password/createNewPasswordScreen.tsx [hunk 29]
```diff
@@ -0,0 +1,122 @@
...
+                            disabled={!!(values.password?.length == 0 || values.confirmNewPassword?.length == 0)}
```
Comment (domsec) [inline packages/client/screens/reset-password/createNewPasswordScreen.tsx:99]: I'm not sure this works based on the attached images (enabled when form fields are empty)

File: packages/client/screens/reset-password/resetPasswordEmailScreen.tsx [hunk 30]
```diff
@@ -0,0 +1,79 @@
...
+                                value={values.email}
+                                onChangeText={handleChange("email")}
+                                placeholder="Email"
+                                testID="email"
+                            />
+                            <FormControl.ErrorMessage>
+                                {errors.email != ERROR_NOP && errors.email}
+                            </FormControl.ErrorMessage>
+                        </FormControl>
+                        <Button style={{ marginTop: 7 }} onPress={submitForm}>
```
Comment (domsec) [inline packages/client/screens/reset-password/resetPasswordEmailScreen.tsx:62]: Disable this button when the email form field is empty.

File: packages/client/navigation/index.tsx [hunk 31]
```diff
@@ -742,6 +835,64 @@ function GuestStackScreen() {
...
                 }}
             />
+            <GuestStack.Screen
+                name="ResetPasswordEmailScreen"
+                component={ResetPasswordEmailScreen}
+                options={{
+                    title: "",
+                    headerTintColor: "#404040",
+                    headerShadowVisible: false,
+                }}
```
Comment (domsec) [inline packages/client/navigation/index.tsx:845]: Add custom back button icon in top nav
```suggestion
                options={({ navigation }) => ({
                    title: "",
                    headerTintColor: "#404040",
                    headerShadowVisible: false,
                    headerLeft: () => BackButton(navigation),
                }}
```

**Review by domsec — APPROVED** [review]
I'm pretty sure I got all the UI specific issues

File: packages/client/screens/reset-password/createNewPasswordScreen.tsx (hunk 10 above)
Comment (alexbara2000) [inline packages/client/screens/reset-password/createNewPasswordScreen.tsx:28]: Next time I suggest doing console.log("here") in order to easily remove it in the future

**Review by alexbara2000 — APPROVED** [review]
good effort, make sure you don't break master

File: packages/api/src/services/users.service.ts (hunk 4 above)
Comment (Andre-Ibrahim) [reply packages/api/src/services/users.service.ts:83]: the update user service requires the current password to be able to update the password which is needed in a forget password use case

File: packages/client/screens/reset-password/createNewPasswordScreen.tsx (hunk 29 above)
Comment (Andre-Ibrahim) [reply packages/client/screens/reset-password/createNewPasswordScreen.tsx:99]: yup changed to the field isDisabled which was set to also control the styling

**Pull request merged**